content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,275 @@
+The container runtime assigns both CPU request and CPU limit to a container.
+Provided the system has free CPU time, containers are guaranteed to be 
+allocated as much CPU as it requests. Containers cannot use more CPU than
+the configured limit i.e. containers CPU usage will be throttled if they
+use more CPU then the specified limit within a given time slice.
+
+Prior to Memory QoS feature, the container runtime only used the memory
+limit and discarded the memory request. If a container uses more memory
+than the configured limit, the Linux Out Of Memory (OOM) killer will kill
+the process in the container that tried to allocate the extra memory.
+
+Let's compare how the container runtime on Linux typically configures memory
+request and limit in cgroups, with and without Memory QoS feature:
+
+* **Memory request**
+
+  The memory request is mainly used by kube-scheduler during (Kubernetes) Pod 
+  scheduling. In cgroup v1, there are no controls to specify the minimum amount
+  of memory the cgroup must always retain. Hence, the container runtime did not
+  use the value of requested memory set in the Pod spec.
+
+  cgroups v2 introduced a `memory.min` setting, used to specify the minimum amount of memory
+  a cgroup must always retain. If the memory usage of a cgroup is within its effective
+  min boundary, the cgroup’s memory won’t be reclaimed under any conditions. If 
+  there's no unprotected reclaimable memory available, the kernel invokes its 
+  OOM killer. Memory QoS maps `memory.min` to `spec.containers[].resources.requests.memory`
+  to ensure the availability of memory for containers in Kubernetes Pods. 
+
+* **Memory limit**
+  
+  The memory limit specifies the memory limit, beyond which if the container tries to allocate 
+  more memory, Linux kernel will terminate it with an OOM (Out of Memory) kill.
+
+  In cgroup v1, `memory.limit_in_bytes` interface is used to set the memory usage limit.
+  However, unlike CPU, it was not possible to apply memory throttling: as soon as a
+  container crossed the memory limit, it would be OOM killed.
+
+  In cgroup v2, `memory.max` is analogous to `memory.limit_in_bytes` in cgroupv1.
+  Memory QoS maps `memory.max` to `spec.containers[].resources.limits.memory` to
+  specify the hard limit for memory usage. If the memory consumption goes above this
+  level, the kernel invokes its OOM Killer.
+
+  cgroup v2 also added `memory.high` configuration . Memory QoS uses `memory.high`
+  to set memory usage throttle limit. If the `memory.high` limit is breached,
+  the offending cgroups are throttled, and the kernel tries to reclaim memory
+  which may avoid an OOM kill.
+
+## How it works
+
+### Cgroup v2 memory controller interfaces & Kubernetes container resources mapping
+
+Memory QoS uses the memory controller of cgroup v2 to guarantee memory resources in
+Kubernetes. cgroupv2 interfaces that this feature: 
+* `memory.max`
+* `memory.min`
+* `memory.high`.
+
+![](./memory-qos-cal.svg)
```

Comment:
nit: we prefer to refer to images using site-relative paths (eg `/blog/2023/03/08/qos-memory-resources/memory-qos-cal.svg`)

nit: we like to set alt attributes to help with accessibility

nit: we recommend using a [figure](https://gohugo.io/content-management/shortcodes/#figure) shortcode for technical figures